---
title: "Homework1"
output:
  html_document:
    df_print: paged
  pdf_document: null
header-includes:
- \usepackage {hyperref}
- \hypersetup {colorlinks = true, linkcolor = blue, urlcolor = blue}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/JHU_Fall_2020/Data_Analysis/Datasets")
```

#### Brian Wiley
### Homework 1
### AS.410.671.81.FA20 Gene Expression Data Analysis and Visualization
Links to question:  [1.](#q1) [2.](#q2) [3.](#q3) [4.](#q4) [5.](#q5) [6.](#q6) [7.](#q7) [8.](#q8) [9.](#q9) [10.](#q10) [11.](#q11)

#### Summary
For this homework, we will be working with a study from Gene Expression Omnibus (GEO) with the accession [GDS2880](https://www.ncbi.nlm.nih.gov/sites/GDSbrowser?acc=GDS2880).  This is an Affymetrix microarray experiment (HGU133A array).  The data researchers were investigating patient matched normal and stage 1 or stage 2 clear cell renal cell carcinoma (cRCC) tumors to provide insight into the molecular pathogenesis of cRCC.  We will be conducting outlier analysis using various methods to identify aberrant samples, followed by missing value imputation to assess the accuracy of two different algorithms.


##### **1.) Download and load the renal cell carcinoma data file into R.  Make sure that the row names are in the correct location (Affymetrix fragment names).  Look at the dimensions and verify that you have 22 arrays and 22,283 probesets. (2pts.)** {#q1}
```{r}
renal.cell <- read.table("renal_cell_carcinoma.txt", header = T, row.names = 1)
dim(renal.cell)
names(renal.cell)

```

##### **2.)Label the header columns of your data frame maintaining the GSM ID, but adding the Normal/Tumor identity. (2pts.)** {#q2}
```{r}
renal.annot <- read.table("renal_carcinoma_annotation.txt", row.names = 1,
                          stringsAsFactors = F)

## to make sure we have to correct order we should always map
## never assume that the annotation is same order as data file
cols <- names(renal.cell)
## a data frame is a list
## https://stackoverflow.com/questions/7070173/convert-data-frame-column-to-a-vector
groups <- renal.annot[cols, ][["V9"]]
new.cols <- paste0(cols, ".", groups)

names(renal.cell) <- new.cols
names(renal.cell) 
```

##### **3.) Identify any outlier samples using the following visual plots:**{#q3}

##### **Correlation plot (heat map)	(2pts.)**
##### **Hierarchical clustering dendrogram (2pts.)**
##### **CV vs. mean plot (2pts.)**
##### **Average correlation plot (2pts.)**
##### **For all plots, make sure you label the points appropriately, title plots, and label axes.  You will also need to provide a legend for the correlation plot.  You can use the gplots for a color gradient, or just use the default colors.** 

Correlation Plot
```{r plotting, fig.height=6, fig.width=6}

## correlation plot
data_cor <- cor(renal.cell, use="pairwise.complete.obs", method="pearson")

library(ggplot2)
library(reshape2)

melt_data_cor <- melt(data_cor)
mid <- (min(melt_data_cor$value)+max(melt_data_cor$value))/2

ggplot(melt_data_cor, aes(x=Var1, y=Var2, fill=value)) +
  geom_tile(color="white") +
  scale_fill_gradient2(low="blue", mid="black", high="yellow",
                       midpoint=mid, 
                       limit=c(min(melt_data_cor$value), 1),
                       name="Pearson\nCorr.") +
  theme(axis.text.x=element_text(angle=90),
        axis.title=element_blank(),
        plot.title=element_text(hjust=0.5)) +
  ggtitle("Rena Cell Carcinoma - Tumor vs. Normal\nCorrelation Plot")

```
From the correlation plot we see there are two samples `GSM146798.Normal` and `GSM146799.Tumor` that show low correlation even within samples of the same group.


Hierarchical clustering dendrogram 
```{r}
dat <- t(renal.cell)
## calculate distances
dat.dist <- dist(dat)
dat.clust <- hclust(dat.dist, method = "complete")

## plot and annotate
plot(dat.clust)
rect(0.5, -800000, 2.5, 300000, border = "red", lwd = 2)
arrows(5, 500000, 3, 350000, col = "blue", lwd = 1.5)
text(5.2, 550000, labels = "Possible Outliers", adj = 0, cex = 0.8, col = "blue")
```
Again these two sample arrays, `GSM146798.Normal` and `GSM146799.Tumor`, that were noticed from the correlation plot are part of their own cluster outside from the two main clusters  We will most likely remove these samples from the analysis.


In addition to the correlation plot and hierarchial clustering plots as their own plots, we can combine these plots using `heatmap.2` function from `gplots` into a dendogram based on the correlation.  This makes is much easier to see the two outluer samples because of the cold blue
created by their hierarchial clustering. `GSM146798.Normal` and `GSM146799.Tumor` have a .98 Pearson correlation coefficient.
```{r 3. Extra, fig.width=7, fig.height=6}

library(gplots) # for heatmap.2 to see combination of heatmap and dendrogram
col <- colorpanel(25, "red", "yellow", "blue")
## b l t r (bottom, left, top, right)
par(mar=c(1,1,1,1)+0.1)
par(oma=c(1,1,1,1))

heatmap.2(data_cor, col = rev(col), trace = "none", margins = c(10,10))

## how correlated are these samples from different groups
data_cor["GSM146798.Normal", "GSM146799.Tumor"]

```


CV vs. mean plot
```{r, fig.width=7}
## get cv and mean for each sample
means = apply(renal.cell, 2, function(x) mean(x, na.rm=T))
cvs = apply(renal.cell, 2, function(x) sd(x)/mean(x))

## plot
plot(means, cvs, xlab = "Mean", ylab = "CV", col="red", pch=20,
     xlim = c(min(means), max(means)*1.1))
title("Renal Cell Carcinoma\nSample CV vs. Mean")
text(means, cvs, labels=names(renal.cell), pos = 4, cex = 0.7)
```
Again we see these 2 samples very far away having a much higher mean and CV then the rest of the samples.


Average correlation plot
```{r, fig.width=7, fig.height=5}
avg.cor <- apply(data_cor, 1, mean)

## conver to df for ggplot
avg.cor <- as.data.frame(avg.cor)
avg.cor$x = row.names(avg.cor)
##https://stackoverflow.com/questions/20041136/avoid-ggplot-sorting-the-x-axis-while-plotting-geom-bar/20041311
avg.cor$x <- factor(avg.cor$x, levels=avg.cor$x)

## plot
ggplot(avg.cor, aes(x = x, y = avg.cor)) +
  geom_point(aes(fill="red"), colour = "black", pch=21, stat = "identity") +
  theme(axis.text.x=element_text(angle=90),
        plot.title=element_text(hjust=0.5),
        axis.title.x=element_blank(),
        legend.position="none",
        # t, r, b, l
        plot.margin = unit(c(0.5,1,0.5,0.5), "cm"),
        #https://stackoverflow.com/questions/14487188/increase-distance-between-text-and-title-on-the-y-axis
        axis.title.y = element_text(margin = margin(t = 0, r = 10, b = 0, l = 0))) +
  ylab("Avg r") +
  ggtitle("Average Correlation Plot")

```
This plot also shows the same two outlier samples from the other 3 plot types.  It also shows another sample, `GSM146799` has pretty low average correlation but I will not remove this same, only `GSM146798.Normal` and `GSM146799.Tumor`.


##### **4.) Install and load the impute library. (1pt.)**{#q4}
```{r, message=FALSE}
library(impute)

```


##### **5.) Remove the outlier samples you identified in the first part of this assignment. (2pts.)**{#q5}
```{r, message=FALSE, fig.width=7, fig.height=6}
drop.cols <- c('GSM146798.Normal', 'GSM146799.Tumor')
renal.cell.rem.out <- renal.cell[, !(names(renal.cell) %in% drop.cols)]

## confirm removed and heatmap correlations with dendrogram
heatmap.2(cor(renal.cell.rem.out, use="pairwise.complete.obs", method="pearson"), 
          col = rev(col), trace = "none", margins = c(10,10))

```


##### **6.) Now we are going to use a couple of transcripts that were determined in this study to be indicative of normal renal function.  The genes we will assess are kininogen 1 (KNG1) and aquaporin 2 (AQP2).  Using either NetAffx or Gene Cards websites (or other resources, if you like), extract the probesets for these two genes.  Hint: KNG1 has two while AQP2 has one.  Then plot a profile plot (expression intensity vs. samples) for each probeset for these two genes.  You may have to convert the data frame row to a vector to plot it.  Do the plots of these genes seem to indicate normal renal function? Explain. (6pts.)**{#q6}
```{r, fig.width=8, fig.height5}
## there were three results for KNG1
c("206054_at", "217512_at", "243562_at") %in% row.names(renal.cell.rem.out)
## '243562_at' is in the U133B Array

## '206672_at' is for AQP2 in U133A
renal.genes <- as.data.frame(t(renal.cell.rem.out[c("206054_at", "217512_at", "206672_at"), ]))

## https://stackoverflow.com/questions/23246070/how-to-use-names-and-rownames-of-a-dataframe-for-the-aes-of-ggplot/23246525
library(reshape2)
renal.genes$sample <- row.names(renal.genes)
renal.genes.final <- melt(renal.genes, id.vars = "sample", value.name = "intensity")
## sanity check
nrow(renal.genes.final) == 3 * dim(renal.cell.rem.out)[[2]]

## again factor to keep order
renal.genes.final$sample <- factor(renal.genes.final$sample, levels = unique(renal.genes.final$sample))
## do log2 transform of intensities for better scale factor
renal.genes.final$intensity <- log2(renal.genes.final$intensity)

## plot
## https://community.rstudio.com/t/ggplot2-change-legend-title-while-controlling-line-types-and-colors/14966/4
lcols = c("blue", "blue", "red")
lty = c(1,2,1)

ggplot(renal.genes.final, aes(x = sample, y = intensity, linetype=variable, color=variable)) +
  geom_line(aes(group=variable)) +
  geom_point(aes(group=variable)) +
  scale_linetype_manual("Genes", labels=c("206054_at (KNG1)", "217512_at (KNG1)", "206672_at (AQP2)"), values=lty) +
  scale_color_manual("Genes", labels=c("206054_at (KNG1)", "217512_at (KNG1)", "206672_at (AQP2)"), values=lcols) +
  theme(axis.text.x=element_text(angle=90),
        plot.title=element_text(hjust=0.5),
        # https://stackoverflow.com/questions/28872875/ggplot2-applying-width-of-line-to-the-legend-key
        legend.key.width = unit(1,"cm"),
        axis.title.x=element_blank()) +
  ggtitle("Gene Profiles for KNG1 & AQP2")

```
It was much easier visually to inspect after log2 transformation.  It is obvious now that these genes have higher expression in normal sample and lower expression in tumor samples alluding to their role required for normal renal function.


##### **7.) We want to assess the accuracy of missing value imputation.  So assign the KNG1 probeset (206054_at) an NA value, only for array GSM146784.  Be sure to first save the original value before replacing it with an NA.  Also cast the data frame to a matrix to run this function. (2pts.)**{#q7}
```{r}
#renal.cell.rem.out <- renal.cell[, !(names(renal.cell) %in% drop.cols)]
sample.GSM146784 <- grep("GSM146784.", names(renal.cell.rem.out))
value <- renal.cell.rem.out["206054_at", sample.GSM146784]
renal.cell.rem.out["206054_at", sample.GSM146784] = NA

renal.cell.rem.out.matrix <- as.matrix(renal.cell.rem.out)

```


##### **8.) Now estimate the missing values in the array using 6 nearest neighbors and Euclidean distance with the impute.knn() function. (2pts.)**{#q8}
```{r}
## assign imputed data
renal.cell.rem.out.matrix.knn <- impute.knn(renal.cell.rem.out.matrix, k=6)

```

##### **9.) Now estimate the missing values in the array using 6 nearest neighbors and Euclidean distance with the `impute.knn()` function. (2pts.)**{#q9}
```{r}
## get inputed value
imputed.value <- renal.cell.rem.out.matrix.knn$data["206054_at", sample.GSM146784]

## get relative error
paste0(round(abs(value-imputed.value)/value*100, 2), "% relative error")

```


##### **10.) Now impute the missing values using the SVD imputation method.  This is in the `pcaMethods` package and the function is called `pca()` with method svdImpute and set `nPcs=9`.  To retrive the output matrix, see the help file. (2pts.)**{#q10}
```{r}
BiocManager::install("pcaMethods")
library(pcaMethods)
?pcaMethods::pca
listPcaMethods()
#renal.cell.rem.out.matrix["206054_at", sample.GSM146784]
## assign imputed data
#renal.cell.rem.out.matrix.svd
renal.cell.rem.out.matrix.svd <- pcaMethods::pca(renal.cell.rem.out.matrix, method="svdImpute", nPcs=9)

## get matrix from the slot (@) completeObs
## confirm dimensions
dim(renal.cell.rem.out.matrix.svd@completeObs)

## get inputed value
imputed.value.svd <- renal.cell.rem.out.matrix.svd@completeObs["206054_at", sample.GSM146784]

## get relative error
paste0(round(abs(value-imputed.value.svd)/value*100, 2), "% relative error")

```

##### **11.) Finally, plot a gene profile plot of the probeset for this gene, where the two different imputed values are represented as different colored points and the actual value is a third point. (6pts.) **{#q10}
```{r}
## assign value back in
renal.cell.rem.out["206054_at", sample.GSM146784] = value
gene <- as.data.frame(t(renal.cell.rem.out["206054_at", ]))
gene$x <- row.names(gene)
gene$x <- factor(gene$x, levels = gene$x)

ggplot(gene, aes(x=x, y=`206054_at`, group=1)) +
  geom_line(color="blue") +
  geom_point(aes(col="blue")) +
  geom_point(aes(x=row.names(gene)[sample.GSM146784], y=imputed.value, col="green")) +
  geom_point(aes(x=row.names(gene)[sample.GSM146784], y=imputed.value.svd, col="red")) +
  scale_color_manual("GSM146784 Values",
                    values = c("blue", "green", "red"),
                    labels = c("actual", "knn imputed", "svd imputed")) +
  theme(axis.text.x=element_text(angle=90),
        plot.title=element_text(hjust=0.5),
        axis.title.x=element_blank()) +
  ylab("206054_at intensities") +
  ggtitle("Gene Profile Plot for 206054_at (KNG1)")
```

