---
title: "Lab1_Brian_Wiley"
output:
  pdf_document:
header-includes:
    - \usepackage {hyperref}
    - \hypersetup {colorlinks = true, linkcolor = blue, urlcolor = blue}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/media/coyote/5838-D2D1/JHU_2021/week1/Datasets")
```

### Brian Wiley
### Lab 1: Exon array data
### AS.410.713.81.SP21 Advanced Genomics and Genetics Analyses
Links to question:  [1.](#q1) [2.](#q2) [3.](#q3) [4.](#q4) [5.](#q5) [6.](#q6) [7.](#q7) [8.](#q8)
<br></br>

#### Summary
This lab introduces concepts for analyzing exon array data.  The dataset that we are using has both exon- and gene-level summaries, separated into 2 distinct data matrices.  The data was taken from a study that was designed to investigate alternative splicing between pediatric patients that have experienced liver rejection after an organ transplant and pediatric patients that have not experienced such rejection.  These are blood samples taken from these patients, so the effect sizes are rather modest.  As a result, we will be a little more liberal in our exon probe filtering methods to retain as much information as possible.
<br></br>

##### **1.) Obtain and load in the exon-rma-sketch.summary.txt, gene-rma-sketch.summary.txt, dabg.summary.txt, and HuEx-1_0-st-v2.na24.hg18.probeset_abbr.csv files into R.** {#q1}
```{r }
map <- read.csv("HuEx-1_0-st-v2.na24.hg18.probeset_abbr.csv", header = T, row.names = 1)
e <- read.table("exon-rma-sketch.summary.txt", skip = 70, row.names = 1, header = T)
g <- read.table("gene-rma-sketch.summary.txt", skip = 70, row.names = 1, header = T)
p <- read.table("dabg.summary.txt", skip = 70, row.names = 1, header = T)
map["2315941",]
rownames(map[map$transcript_cluster_id=="2315918",])
e[rownames(map[map$transcript_cluster_id=="2315918",]),]
p["2315941",]
```
```{r}
map[map$transcript_cluster_id=="2315918",]
g[, ]

```
<br></br>
<br></br>

##### **2.)Continue to follow code in the lecture, up to the line of: ** {#q2}
 `#you should filter probes based on guidelinesâ€¦.`

```{r }
r <- c(1,0,1,1,1,0,1,1,1,0,1,0,0,0,0,1,1,0,0,0,0,0,0,0,0,1)
x <- intersect(dimnames(e)[[1]], dimnames(map)[[1]])
e <- e[x,]
p <- p[x,]
map <- map[x,]
u <- unique(as.character(map$transcript_cluster_id))
u <- intersect(u,dimnames(g)[[1]])

```

<br></br>
<br></br>
##### **3.) Filter probes in the exon matrix using the detection above background (DAGB) p-values.  If >50% of the subjects in BOTH groups have a p-value > .05, remove this exon probe.  Note that this is a little more liberal than the thresholds that we explained in the lecture.  The effects are small for this dataset, so this is why we are being a little less conservative with our filtering.  How many exon probes remain after this filtering?  Subset the exon matrix by those probes that pass the filter. ** {#q3}

<br></br>
<br></br>
163972 of the 287220 probes from the DABG did NOT have both groups with 50% of p-values > 0.05 
and passed the filter.
```{r }
## use ragged array to get percent of p-values above 0.05 for each group
percent.above.p05 <- apply(p, 1, function(x) tapply(x, r, function(y) sum(y > 0.05)/length(y)))
both.greater.than50 <- apply(t(percent.above.p05), 1, function(x) all(x > 0.5))
#both.greater.than50[row.names(map.fail[map.fail$transcript_cluster_id == clust, ])]
## how many kept
sum(!both.greater.than50)

## are exon probes for 'dabg' and 'exons' in same order
all(row.names(p) == row.names(e))

## since 'dabg' is 287329 in length and 'e' is 287200 in length cannot just use boolean array
#exon.probes.keep <- names(both.greater.than50)[which(!both.greater.than50)]
# sanity
#length(exon.probes.keep) == sum(!both.greater.than50)
#exons.pass <- e[exon.probes.keep, ]
exons.pass = e[!both.greater.than50, ]
dim(exons.pass)
```


##### **4.) Go back to the annotation table (i.e. map) and subset it by the shorter set of exon probes that you will now be using.  Get the unique transcript cluster IDs from this new annotation table and intersect them with the gene matrix.  You should now have a shorter u variable.  How many unique transcript cluster IDs do you have? ** {#q4}

<br></br>
<br></br>
There are now 16659 unique transcript clusters.
```{r }
#sanity
all(row.names(map) == row.names(e))

map.pass <- map[!both.greater.than50, ]
map.fail <- map[both.greater.than50, ]
new.u <- unique(as.character(map.pass$transcript_cluster_id))
new.u <- intersect(new.u,dimnames(g)[[1]])
cat("New u length: ", length(new.u), "\n")
cat("Old u length: ", length(u))
```


##### **5.) Now, using the code in the lecture, write a loop that loops through all values of the variable u (unique transcript cluster ID).  In each loop iteration, you need to:** {#q5}
<br></br>    
<br></br>          
        a. Get the correct mapping between the ith transcript cluster ID and the exon probe IDs
        b. Subset the gene and exon matrices by the appropriate transcript cluster ID or exon probe IDs
        c. Call the exon.ni function
        d. Store the output p-values, test statistics, and CIs from this function for each transcript cluster ID (and exon probe IDs)
For this loop, you also need to add an if statement to only conduct this calculation when you have at least 2 rows in the d.exon matrix (i.e. at least 2 exon probe IDs for a single transcript cluster ID).  This loop will take up to an hour to run (depending on your computer speed), so plan accordingly.** 

```{r}
t.two <- function(x,sam,v=F) {
  x <- as.numeric(x)
  out <- t.test(as.numeric(x[sam]),as.numeric(x[!sam]),alternative="two.sided",var.equal=v)
  o <- as.numeric(c(out$statistic,out$p.value,out$conf.int[1],out$conf.int[2]))
  names(o) <- c("test_statistic","pv","lower_ci","upper_ci")
  return(o)
}
ni <- e["2315941",]
g
exon.ni <- function(genex, exonx, rx) {
  ni <- t(t(exonx)-genex)
  ttest <- t(apply(ni,1,t.two,sam=as.logical(rx),v=F))
  return(ttest)
}


ex <- row.names(map[map$transcript_cluster_id == new.u[1], ])
d.exon <- e[ex, ]
d.gene <- g[new.u[1],]
ni.out <- exon.ni(genex=as.numeric(d.gene), exonx=d.exon, rx=r)
class(ni.out)


## keep track of single exon clusters for fun
single.exons <- c()
j=1

## empty matrix for summary statistics
## need 5 for adding the cluster id column in loop
exon.ni.mat <- matrix(,ncol = 5)


##timeit
ptm <- proc.time()
for (clust in new.u) {
  ex <- row.names(map.pass[map.pass$transcript_cluster_id == clust, ])
  #row.names(map[map$transcript_cluster_id == clust, ])
  #row.names(map.fail[map.fail$transcript_cluster_id == clust, ])
  if (length(ex) < 2) {
    print(clust)
    single.exons[j] <- clust
    j = j + 1
    next
  }
  d.exon <- exons.pass[ex, ]
  d.gene <- g[clust, ]
  ni.out <- exon.ni(genex=as.numeric(d.gene), exonx=d.exon, rx=r)
  ## add cluster id column
  ni.out.with.cluster <- cbind(ni.out, as.numeric(clust))
  exon.ni.mat <- rbind(exon.ni.mat, ni.out.with.cluster)
}
total.time <- proc.time() - ptm
## remove empty first row
exon.ni.mat <- exon.ni.mat[-1,]
colnames(exon.ni.mat) <- c("test_statistic","pv","lower_ci","upper_ci","cluster")
## how many minutes
total.time/60
```


##### **6) Sort the matrix where you stored all of the summary statistics by p-value and get the transcript cluster ID with the lowest p-value.** {#q6}
```{r}
head(exon.ni.mat)
dim(exon.ni.mat)
sorted <- exon.ni.mat[order(exon.ni.mat[,'pv']),]
head(sorted)
cluster <- sorted[,'cluster'][1]
lowest.p.exon <- row.names(sorted)[1]

```


##### **7) Sort the matrix where you stored all of the summary statistics by p-value and get the transcript cluster ID with the lowest p-value.** {#q7}

```{r, fig.width=8, fig.height=8}
plot.exons <- function(exonx,genex,rx,ti) {
  ## for legend inside plot
  n <- nrow(exonx)
  rr <- rx
  rx <- rep(rx,nrow(exonx))
  rx[rx==1] <- "A"
  rx[rx==0] <- "B"
  rx <- as.factor(rx)
  ni <- t(t(exonx)-genex)
  exonx <- as.data.frame(t(ni))
  ex.stack <- stack(exonx)
  d <- data.frame(ex.stack,rx)
  names(d) <- c("exon_values","exon_id","class")
  d$exon_id <- as.factor(d$exon_id)
  d$class <- as.factor(d$class)
  genex.title <- as.character(map[match(ti,as.character(map$transcript_cluster_id)),"gene_assignment"])
  
  # plot(c(.5,(ncol(exonx)+.5)),range(d[,1]),type="n",axes=F,xlab="",ylab="")
  # boxplot(exon_values~exon_id,add=T,subset=d$class=="A",d,col="salmon",border='red',cex.axis=.75,las=2,ylab='Log2 normalized intensity',main=paste("Gene ID:",ti,"\n",genex.title),boxwex=0.4)
  # boxplot(exon_values~exon_id,subset=d$class=="B",d,add=T,col="green",border='darkgreen',axes=F,boxwex=0.4,
  # at=c(1:ncol(exonx))+0.1)
  
  
  library(ggplot2)
  ## plot significant value asterisks
  library(ggsignif)
  ##https://stackoverflow.com/questions/55674679/how-to-add-significance-bar-between-subgroups-of-box-plot
  p.values <- t(apply(ni,1,t.two,sam=as.logical(rr),v=F))[,'pv']
  labels <- symnum(p.values, corr = FALSE, cutpoints = c(0,  .001,.01,.05, 1), symbols = c("***","**","*",""))
  y.values <- sapply(split(d, d$exon_id), 
         function(x){max(sapply(split(x, x$class), 
                                function(xx){boxplot(xx$exon_values, plot=F)$stats[5, ]}))})+.7
  n.s <- p.values > .05
  
  gg <- ggplot(d, aes(x=exon_id, y=exon_values, fill = class, color = class)) +
    geom_boxplot() +
    scale_fill_manual(values=c("salmon", "green", "black"), labels = c("disease", "normal")) +
    scale_color_manual(values=c("red", "darkgreen", "black"), labels = c("disease", "normal")) +
    ylab(expression(paste("Log"[2], " normalized intensity"))) +
    ggtitle(paste("Gene ID:",ti,"\n",genex.title)) +
    theme(axis.text.x=element_text(angle=90),
          plot.title = element_text(hjust = 0.5)) +
    geom_signif(y_position = y.values[!n.s], annotations = labels[!n.s],  xmin = unique(d$exon_id)[!n.s], xmax = unique(d$exon_id)[!n.s], tip_length = 0, color = "black", textsize=6, show.legend = T)  +
    annotate("text", x=n-2, y=max(d$exon_values), label = "
  * p < .05
 ** p < .01
*** p < .001",
             family = "DejaVu Sans Mono")
    
  gg

  
  ##another way to add signficant asterisks
  # library(ggpubr)
  # stat.test <- compare_means(exon_values ~ class, data = d, group.by = "exon_id",
  #                            method = "t.test")
  # 
  # bp <- ggboxplot(d, x = "exon_id", y = "exon_values",
  #               fill = "class", color = "class",
  #               add.params = list(group = "class"), title = paste("Gene ID:",ti,"\n",genex.title))
  # bp + stat_pvalue_manual(
  #   stat.test[c(1,2,3,4,5,8)], x = "exon_id", y.values[stat.test$p<.05],
  #   label = "p.signif", hide.ns = T,
  #   position = position_dodge(0.8)) +
  #   theme(axis.text.x=element_text(angle=90),
  #           plot.title = element_text(hjust = 0.5))
  
}

ex <- row.names(map[map$transcript_cluster_id == cluster, ])
## all exons and not just those that passed DABG < 50% of samples
d.exon <- e[ex,]
d.gene <- g[as.character(cluster),]
plot.exons(exonx=d.exon,genex=as.numeric(d.gene),rx=r,ti=cluster)




```
